### PR TITLE
feat(internal): add MatchMedia component and extend useMedia hook

### PR DESCRIPTION
**Pull request by joshblack**
This PR is a proposal that extends the `useMedia` hook with a `MatchMedia` component to help with emulation of features for development and docs.

_Note: this is only if DevTools does not provide an appropriate alternative._

This may be helpful in documentation scenarios in order to demonstrate the capabilities of a component without a viewer of the docs having that feature turned on (like in [this scenario](https://github.com/primer/react/pull/2297#discussion_r970317953))

It also updates the `useMedia` hook to support this component and adds support for `addEventListener`, where supported.

---

Would love to hear what you all think! Just am opening this up as a suggestion/conversation starter to gauge interest and usefulness

### CODE REVIEW

**Comment by joshblack** [comment]
FYI @colebemis @siddharthkp if you're interested!

**Comment by colebemis** [comment]
Should this replace the existing `useMedia` hook or does it serve a different purpose? https://github.com/primer/react/blob/main/src/hooks/useMedia.ts

**Comment by joshblack** [comment]
@colebemis ah sorry, I was looking for this under `utils` and completely missed the `hooks` folder 🤦 

You're dead on, they have the same purpose so I'll update the existing file for this PR and remove the one added for `matchMedia`!

**Comment by joshblack** [comment]
@colebemis thanks again for the heads up 🙏 Just pushed the changes and updated the description

File: src/utils/environment.js [hunk 1]
```diff
@@ -0,0 +1,6 @@
+/**
+ * Indicate whether current execution environment can access the DOM.
+ *
+ * @see https://github.com/facebook/fbjs/blob/4d1751311d3f67af2dcce2e40df8512a23c7b9c6/packages/fbjs/src/core/ExecutionEnvironment.js#L12
+ */
+export const canUseDOM = !!(typeof window !== 'undefined' && window.document && window.document.createElement)
```
Comment (colebemis) [inline src/utils/environment.js:6]: This will be handy 😍

File: src/hooks/useMedia.tsx [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+ *
+ * This should be used for development and documentation only in situations
+ * where devtools cannot emulate this feature
+ *
+ * @example
+ * <MatchMedia features={{ "(pointer: coarse)": true}}>
+ *   <Children />
+ * </MatchMedia>
+ */
+export function MatchMedia({children, features = defaultFeatures}: MatchMediaProps) {
```
Comment (colebemis) [inline src/hooks/useMedia.tsx:121]: Nice! This will be so helpful for docs 👍

**Review by colebemis — APPROVED** [review]
Love it 💖

File: src/hooks/useMedia.tsx (hunk 2 above)
Comment (broccolinisoup) [reply src/hooks/useMedia.tsx:121]: OMG this is what I needed! 😍  THANK YOU SO MUCH! @joshblack

**Review by broccolinisoup — APPROVED** [review]
Looks awesome 🔥  Looking forward to trying out when it is merged 🚀

File: src/hooks/useMedia.tsx [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+type MatchMediaProps = {
+  children: React.ReactNode
+  features?: MediaQueryFeatures
+}
+
+const defaultFeatures = {}
+
+/**
+ * Use `MatchMedia` to emulate media conditions by passing in feature
+ * queries to the `features` prop. If a component uses `useMatchMedia` with the
```
Comment (joshblack) [inline src/hooks/useMedia.tsx:109]:
```suggestion
 * queries to the `features` prop. If a component uses `useMedia` with the
```

**Pull request merged**